### PR TITLE
test: add package compatibility tests using publint and are the types wrong

### PR DESCRIPTION
**Pull request by alexprudhomme**

### CODE REVIEW

File: scripts/ci/package-compatibility.mjs [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import {publint} from 'publint';
+
+const {messages: atomicMessages} = await publint({
```
Comment (alexprudhomme) [inline scripts/ci/package-compatibility.mjs:3]: This is the file where we want to test whether or not are packages are compatible using [publint](https://publint.dev/) and [attw](https://arethetypeswrong.github.io/)

## Publint
Publint should be easy since they have a js library available. My code here should cover most of our case. We still have warnings for now so only fail if errors. We could add atomic-hosted-page, atomic-angular, bueno, auth, headless-react and more if we want but those three are fine for now imo.

## ATTW 
This one should be a bit harder since there is no js library. You only have a cli tool. So you would have to get the output from this https://www.npmjs.com/package/@arethetypeswrong/cli and compare it to what we want to have. Right now we have issues in CJS, it is fine for now and node10 is meant to fail.

Comment (alexprudhomme) [reply scripts/ci/package-compatibility.mjs:3]: Types in atomic are meant to be broken. this is what attw should look like: 

atomic-react
"@coveo/atomic-react"

node10: 🟢
node16 (from CJS): 👺 Masquerading as ESM
node16 (from ESM): 🟢 (ESM)
bundler: 🟢

***********************************

"@coveo/atomic-react/commerce"

node10: 💀 Resolution failed
node16 (from CJS): 👺 Masquerading as ESM
node16 (from ESM): 🟢 (ESM)
bundler: 🟢

***********************************

"@coveo/atomic-react/recommendation"

node10: 💀 Resolution failed
node16 (from CJS): 👺 Masquerading as ESM
node16 (from ESM): 🟢 (ESM)
bundler: 🟢

***********************************

atomic
"@coveo/atomic/loader"

node10: 🟢
node16 (from CJS): 🥴 Internal resolution error
node16 (from ESM): 🥴 Internal resolution error
🚭 Unexpected module syntax
bundler: 🟢

***********************************

"@coveo/atomic"

node10: 🟢
node16 (from CJS): ⚠️ ESM (dynamic import only)
🚭 Unexpected module syntax
🥴 Internal resolution error
node16 (from ESM): 🥴 Internal resolution error
bundler: 🟢

***********************************

"@coveo/atomic/themes/*"

node10: (wildcard)
node16 (from CJS): (wildcard)
node16 (from ESM): (wildcard)
bundler: (wildcard)

***********************************

"@coveo/atomic/assets/*"

node10: (wildcard)
node16 (from CJS): (wildcard)
node16 (from ESM): (wildcard)
bundler: (wildcard)

***********************************

"@coveo/atomic/lang/*"

node10: (wildcard)
node16 (from CJS): (wildcard)
node16 (from ESM): (wildcard)
bundler: (wildcard)

***********************************

headless
"@coveo/headless"

node10: 🟢
node16 (from CJS): 👺 Masquerading as ESM
node16 (from ESM): 🟢 (ESM)
bundler: 🟢

***********************************

"@coveo/headless/case-assist"

node10: 💀 Resolution failed
node16 (from CJS): 👺 Masquerading as ESM
node16 (from ESM): 🟢 (ESM)
bundler: 🟢

***********************************

"@coveo/headless/commerce"

node10: 💀 Resolution failed
node16 (from CJS): 👺 Masquerading as ESM
node16 (from ESM): 🟢 (ESM)
bundler: 🟢

***********************************

"@coveo/headless/insight"

node10: 💀 Resolution failed
node16 (from CJS): 👺 Masquerading as ESM
node16 (from ESM): 🟢 (ESM)
bundler: 🟢

***********************************

"@coveo/headless/recommendation"

node10: 💀 Resolution failed
node16 (from CJS): 👺 Masquerading as ESM
node16 (from ESM): 🟢 (ESM)
bundler: 🟢

***********************************

"@coveo/headless/ssr"

node10: 💀 Resolution failed
node16 (from CJS): 👺 Masquerading as ESM
node16 (from ESM): 🟢 (ESM)
bundler: 🟢

***********************************

"@coveo/headless/ssr-commerce"

node10: 💀 Resolution failed
node16 (from CJS): 👺 Masquerading as ESM
node16 (from ESM): 🟢 (ESM)
bundler: 🟢

***********************************

Comment (fbeaudoincoveo) [reply scripts/ci/package-compatibility.mjs]: After some tinkering around, I decided not to do the attw check.

The attw CLI is not flexible enough for our purposes. While it's possible to specify a set of rules to ignore, it's not possible to specify for instance that you want to ignore "Resolution failed" only in node10.

I tried to output the result of the command in JSON, which we could then parse in order to only raise process.exit(1) when a "valid" error is detected, but the way the CLI is made, it will process.exit(1) when any error is detected, which means I have to wrap the command in a try-catch in order to run it over all of our packages. When I do this, the JSON stdout gets truncated (even if I write it to a file). In a last desperate attempt, I tried to echo the output of the command directly to a file, but that doesn't work either: the JSON gets truncated.

A possible solution would be to output as ascii (which, at least for now, doesn't get truncated because it's short enough), and compare to an expected ascii output... but this would be very brittle and would, IMO, waste more time than it would save.

In short, I don't think it's worth spending any more time trying to integrate an attw scan in our CI. It's a tool that we can use on our own, but it's ill-adapted to CI use.

**Review by alexprudhomme — COMMENTED** [review]
I was able to make a script with the [zx](https://www.npmjs.com/package/zx) library to test attw as well. In this one, I am testing headless and filtering out the FalseESM errors. Let me know if you think we should add this.

```
import {$} from 'zx';

interface Problem {
  kind: string;
  resolutionKind?: string;
}

await $({
  nothrow: true,
})`npx attw --pack packages/headless --format json > .attw.json`;
const output = await $`cat .attw.json`;
await $`rm .attw.json`;

const json = JSON.parse(output.stdout);

json.analysis.problems = json.analysis.problems.filter(
  (problem: Problem) => problem.kind !== 'FalseESM'
);

if (!json.analysis.problems || json.analysis.problems.length === 0) {
  console.log('No errors found');
  process.exit(0);
}

if (
  json.analysis.problems.every(
    (problem: Problem) => problem.resolutionKind === 'node10 '
  )
) {
  console.log("Only found node10 or problems, which we don't care about");
  process.exit(0);
}

if (
  json.analysis.problems.every((problem: Problem) => problem.kind === 'warning')
)
  console.log('Errors found');
console.log(json.analysis.problems);
process.exit(1);

```

**Comment by fbeaudoincoveo** [comment]
> I was able to make a script with the [zx](https://www.npmjs.com/package/zx) library to test attw as well. In this one, I am testing headless and filtering out the FalseESM errors. Let me know if you think we should add this.

@alexprudhomme as discussed in person, I don't think we should run attw checks in our CI until our packages are clean enough that we aren't forced to filter out "false" positives all over the place in the attw report. Hopefully by then the attw CLI will have evolved to be more CI-friendly, or better yet they'll have released an npm library that does what we need.

**Review by louis-bompart — APPROVED** [review]
